src/program/top/top.lua

```diff
@@ -9,23 +9,159 @@ local shm = require("core.shm")
 local counter = require("core.counter")
 local S = require("syscall")
 local histogram = require("core.histogram")
+local json = require("lib.json")
+local macaddress = require("lib.macaddress")
 local usage = require("program.top.README_inc")
 
 local long_opts = {
-   help = "h"
+   help = "h", counters = "c", yang = "y"
 }
 
 function clearterm () io.write('\027[2J') end
 
 function run (args)
    local opt = {}
+   local object = nil
+   local yang = false
    function opt.h (arg) print(usage) main.exit(1) end
-   args = lib.dogetopt(args, opt, "h", long_opts)
+   function opt.c (arg) object = arg              end
+   function opt.y ()    yang = true               end
+   args = lib.dogetopt(args, opt, "hc:y", long_opts)
 
    if #args > 1 then print(usage) main.exit(1) end
-   local target_pid = args[1]
+   local target_pid = select_snabb_instance(args[1])
 
-   local instance_tree = "//"..(select_snabb_instance(target_pid))
+   if     object then list_counters(target_pid, object)
+   elseif yang   then dump_yang(target_pid)
+   else               top(target_pid) end
+   ordered_exit(0)
+end
+
+function select_snabb_instance (pid)
+   local instances = shm.children("//")
+   if pid then
+      -- Try to use given pid
+      for _, instance in ipairs(instances) do
+         if instance == pid then return pid end
+      end
+      print("No such Snabb instance: "..pid)
+   elseif #instances == 2 then
+      -- Two means one is us, so we pick the other.
+      local own_pid = tostring(S.getpid())
+      if instances[1] == own_pid then return instances[2]
+      else                            return instances[1] end
+   elseif #instances == 1 then print("No Snabb instance found.")
+   else print("Multple Snabb instances found. Select one.") end
+   ordered_exit(1)
+end
```

Comment:
I have a similar issue with my snabbddos-influxdb script and a rather similar approach - https://github.com/SnabbDDoS/snabbddos-influxdb/blob/master/snabbddos-influxdb.py#L33-38

For SnabbDDoS I imagine the proper solution is to have a startup script that wraps around the snabb instance itself, catches the PID of that and then launches monitoring stuff around it with the explicit PID.

As for the stale process directories in /var/run/snabb I think Snabb per default should clean up after itself. That's my gut feeling at least. If someone wants to look at the data after Snabb has finished it's better to have an explicit option for that than the other way around.